### PR TITLE
Convert initialization and undelegation parameters to timestamps

### LINKED ISSUE #1501: Represent initialization and undelegation period in TokenStaking contract as timestamps
As per #1490, customer promise for deposits is expressed in months (timestamps) and token staking periods are represented in blocks. We should update the initialization and undelegation period to use timestamps as well as make sure the rest of the code in `TokenStaking` contract is aligned with timestamps. 

As a part of this issue, we will have to modify KEEP token dashboard.

### PR DESCRIPTION
**Pull request by ngrinkevich**
Refs #1501

Updating for consistency reasons, fwiw initialization and undelegation params don't bring security risks of miner manipulation in this case.

### CODE REVIEW

File: solidity/test/TestTokenStake.js [hunk 1]
```diff
@@ -562,7 +565,7 @@ contract('TokenStaking', function(accounts) {
         operatorOne, operatorContract, {from: authorizer}
       )
   
-      await mineBlocks(initializationPeriod - 1)
+      await increaseTime(initializationPeriod - 1)
```
Comment (pdyraga) [inline solidity/test/TestTokenStake.js]: I think this one should be `increaseTime(initializationPeriod)` to cover the corner case.

File: solidity/test/TestTokenStake.js [hunk 2]
```diff
@@ -574,7 +577,7 @@ contract('TokenStaking', function(accounts) {
   
     it("should report no active stake for not authorized operator contract", async () => {
       await delegate(operatorOne, stakingAmount)
-      await mineBlocks(initializationPeriod);
+      await increaseTime(initializationPeriod);
```
Comment (pdyraga) [inline solidity/test/TestTokenStake.js]: Shouldn't we have +1 here?

File: solidity/test/TestTokenStake.js [hunk 3]
```diff
@@ -643,7 +646,7 @@ contract('TokenStaking', function(accounts) {
         operatorOne, operatorContract, {from: authorizer}
       )
   
-      await mineBlocks(initializationPeriod - 1);
+      await increaseTime(initializationPeriod - 1);
```
Comment (pdyraga) [inline solidity/test/TestTokenStake.js]: Or just `increaseTime(initializationPeriod)` ?

File: solidity/test/TestTokenStake.js [hunk 4]
```diff
@@ -656,7 +659,7 @@ contract('TokenStaking', function(accounts) {
     it("should report no eligible stake for not authorized operator contract", async () => {
       await delegate(operatorOne, stakingAmount)
   
-      await mineBlocks(initializationPeriod);
+      await increaseTime(initializationPeriod);
```
Comment (pdyraga) [inline solidity/test/TestTokenStake.js]: Don't we need to `+1` here?

File: solidity/test/token_grant/TestTokenGrantStake.js [hunk 5]
```diff
@@ -153,7 +151,7 @@ contract('TokenGrant/Stake', function(accounts) {
   it("should allow to cancel delegation just before initialization period is over", async () => {
     await delegate(grantee, operatorOne, grantAmount);
     
-    await mineBlocks(initializationPeriod - 1);
+    await increaseTime(initializationPeriod - 1);
```
Comment (pdyraga) [inline solidity/test/token_grant/TestTokenGrantStake.js]: Or `increaseTime(initializationPeriod)`?

File: solidity/test/token_grant/TestTokenGrantStake.js [hunk 6]
```diff
@@ -184,10 +182,10 @@ contract('TokenGrant/Stake', function(accounts) {
   it("should not allow to recover stake before undelegation period is over", async () => {
     await delegate(grantee, operatorOne, grantAmount);
 
-    await mineBlocks(initializationPeriod);
+    await increaseTime(initializationPeriod + 1);
     await grantContract.undelegate(operatorOne, {from: grantee});
 
-    await mineBlocks(undelegationPeriod - 1);
+    await increaseTime(undelegationPeriod - 1);
```
Comment (pdyraga) [inline solidity/test/token_grant/TestTokenGrantStake.js]: Shouldn't we have just `increaseTime(undelegationPeriod)` here?

File: solidity/contracts/TokenStaking.sol [hunk 7]
```diff
@@ -89,7 +89,7 @@ contract TokenStaking is StakeDelegatable {
         token.safeTransferFrom(_from, address(this), _value);
 
         operators[operator] = Operator(
-            OperatorParams.pack(_value, block.number, 0),
+            OperatorParams.pack(_value, block.timestamp, 0),
```
[the remark this answers is not in the document]
Comment (ngrinkevich) [reply solidity/contracts/TokenStaking.sol]: I think it's disabled globally already in `.soliumrc.json`

**Comment by pdyraga** [comment]
This one does not close #1501, we need to wait for #1513 to do that.

File: solidity/test/token_stake/TestTokenStake.js [hunk 8]
```diff
@@ -147,9 +148,10 @@ contract('TokenStaking', function(accounts) {
   it("should allow to cancel delegation just before initialization period is over", async () => {
     let ownerStartBalance = await token.balanceOf.call(ownerOne);
     
-    await delegate(operatorOne, stakingAmount);
+    let tx = await delegate(operatorOne, stakingAmount)
+    let createdAt = (await web3.eth.getBlock(tx.receipt.blockNumber)).timestamp
 
-    await increaseTime(initializationPeriod);
+    await increaseTimeTo(createdAt + initializationPeriod - 1)
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]: Why we are `-1` here? [Per the previous comment](https://github.com/keep-network/keep-core/pull/1512/files/d339004f82283c04495c8638676b7e87b08fe6c2#r397880959):

> *just before initialization period is over* is at `increaseTime(initializationPeriod)`

File: solidity/test/token_stake/TestTokenStake.js [hunk 9]
```diff
@@ -651,13 +677,12 @@ contract('TokenStaking', function(accounts) {
...
-      await delegate(operatorOne, stakingAmount)
+      let tx = await delegate(operatorOne, stakingAmount)
+      let createdAt = (await web3.eth.getBlock(tx.receipt.blockNumber)).timestamp
       await stakingContract.authorizeOperatorContract(
         operatorOne, operatorContract, {from: authorizer}
       )
-  
-      await increaseTime(initializationPeriod);
-  
+      await increaseTimeTo(createdAt + initializationPeriod - 1)
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]: Why `-1`? I think this one should be `await increaseTimeTo(createdAt + initializationPeriod)`

File: solidity/test/token_grant/TestTokenGrantStake.js [hunk 10]
```diff
@@ -1,6 +1,4 @@
-
```
Comment (pdyraga) [inline solidity/test/token_grant/TestTokenGrantStake.js]: All `increaseTime` values look good to me but shouldn't we use `increaseTimeTo` in this file as well?

File: solidity/test/token_stake/TestTokenStake.js [hunk 11]
```diff
@@ -267,52 +273,68 @@ contract('TokenStaking', function(accounts) {
...
     })
 
-    it("should not permit undelegating at the block before initialization " + 
+    it("should not permit undelegating at the time before initialization " + 
     "period passed", async () => {
-      await delegate(operatorOne, stakingAmount)
+      let tx = await delegate(operatorOne, stakingAmount)
+      let createdAt = (await web3.eth.getBlock(tx.receipt.blockNumber)).timestamp
+
+      await increaseTimeTo(createdAt + initializationPeriod - 1)
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]: [This](https://github.com/keep-network/keep-core/pull/1512/files/d339004f82283c04495c8638676b7e87b08fe6c2#r397889897) comment still applies, I think we should have `await increaseTimeTo(createdAt + initializationPeriod)` here.

File: solidity/test/token_stake/TestTokenStake.js [hunk 12]
```diff
@@ -177,12 +181,13 @@ contract('TokenStaking', function(accounts) {
...
-    await mineBlocks(initializationPeriod);
-    await stakingContract.undelegate(operatorOne, {from: ownerOne});
+    let tx = await delegate(operatorOne, stakingAmount)
+    let createdAt = (await web3.eth.getBlock(tx.receipt.blockNumber)).timestamp
 
-    await mineBlocks(undelegationPeriod - 1);
+    await increaseTimeTo(createdAt + initializationPeriod + 1)
+    tx = await stakingContract.undelegate(operatorOne, {from: ownerOne});
+    let undelegatedAt = await web3.eth.getBlock(tx.receipt.blockNumber);
+    await increaseTimeTo(undelegatedAt.timestamp + undelegationPeriod);
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]:
```
  1) Contract: TokenStaking
       should not allow to recover stake before undelegation period is over:
     AssertionError: Expected throw not received
      at _callee$ (test/helpers/expectThrowWithMessage.js:9:10)
      at tryCatch (node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:65:40)
      at Generator.invoke [as _invoke] (node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:303:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:117:21)
      at step (test/helpers/expectThrowWithMessage.js:7:191)
      at /Users/piotr/go/src/github.com/keep-network/keep-core/solidity/test/helpers/expectThrowWithMessage.js:7:361
      at processTicksAndRejections (internal/process/task_queues.js:86:5)

```

locally :(

File: solidity/test/token_stake/TestTokenStake.js [hunk 13]
```diff
@@ -557,12 +592,13 @@ contract('TokenStaking', function(accounts) {
...
     it("should report no active stake before initialization period is over", async () => {
-      await delegate(operatorOne, stakingAmount)
+      let tx = await delegate(operatorOne, stakingAmount)
+      let createdAt = (await web3.eth.getBlock(tx.receipt.blockNumber)).timestamp
       await stakingContract.authorizeOperatorContract(
         operatorOne, operatorContract, {from: authorizer}
       )
   
-      await mineBlocks(initializationPeriod - 1)
+      await increaseTimeTo(createdAt + initializationPeriod - 1)
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]: [This](https://github.com/keep-network/keep-core/pull/1512/files/d339004f82283c04495c8638676b7e87b08fe6c2#r397994846) comment still applies. I think we should have `await increaseTimeTo(createdAt + initializationPeriod)` here. This is just before the initialization period is over.

File: solidity/test/token_stake/TestTokenStake.js [hunk 14]
```diff
@@ -305,6 +305,16 @@ contract('TokenStaking', function(accounts) {
       )
     })
 
+    it("should not permit undelegating at the time when initialization " +
```
Comment (pdyraga) [inline solidity/test/token_stake/TestTokenStake.js]: We should permit undelegating when initialization period passed and we should not permit before. And we have test cases for those scenarios declared just before this one:
- should permit undelegating at the time when initialization period passed
- should not permit undelegating at the time before initialization period passed

File: solidity/test/token_stake/TestTokenStake.js (hunk 12 above)
Comment (ngrinkevich) [reply solidity/test/token_stake/TestTokenStake.js]: yeah try to run it again and you'll get different results 😅 I think we have to keep +1 / -1 seconds adjustments, maybe even switch to minutes?

File: solidity/test/token_stake/TestTokenStake.js (hunk 14 above)
Comment (ngrinkevich) [reply solidity/test/token_stake/TestTokenStake.js]: ah sorry I got it wrong then, do you remember what corner case did you mean in the previous review?

File: solidity/test/token_stake/TestTokenStake.js (hunk 12 above)
Comment (pdyraga) [reply solidity/test/token_stake/TestTokenStake.js]: Yeah... To do not lose any clarity about `-1 / +1` I think we should introduce a `const timeRoundMargin` set to one minute or so and subtract this margin from all tests stressing *before* cases. For example, *should not permit undelegating at the time before initialization period passed* would switch from `await increaseTimeTo(createdAt + initializationPeriod)` to `await increaseTimeTo(createdAt + initializationPeriod - timeRoundMargin)`. All *after* cases should stay as-is. This way it's clear `createdAt + initializationPeriod` is the last time when the undelegation is allowed but we need to use `timeRoundMargin` to avoid randomly failing tests. We also need to document `timeRoundMargin` well. What do you think?

File: solidity/test/token_stake/TestTokenStake.js (hunk 14 above)
Comment (pdyraga) [reply solidity/test/token_stake/TestTokenStake.js]: That comment was all about removing `-1 / +1` from `increaseTime / increaseTimeTo` so that both mentioned tests use margin values. I looked at the code and it's all correct now so I think we can just drop the test from line 308.

**Comment by pdyraga** [comment]
Three pieces left:
- [Redundant test at line 308](https://github.com/keep-network/keep-core/pull/1512#discussion_r399548953)
- [Tests failing from time to time](https://github.com/keep-network/keep-core/pull/1512#discussion_r399545609)
- [decision on `increaseTime -> increaseTimeTo` in `TestTokenGrantStake`](https://github.com/keep-network/keep-core/pull/1512#discussion_r399540910)
- nitpick: [not sure if the maximum allowed date comment is correct](https://github.com/keep-network/keep-core/pull/1512#discussion_r397852160)

File: solidity/test/token_stake/TestTokenStake.js (hunk 12 above)
Comment (ngrinkevich) [reply solidity/test/token_stake/TestTokenStake.js]: >We also need to document timeRoundMargin well. What do you think

yep `timeRoundMargin` constant is a good idea 👌I'll update
